esmvaltool/diag_scripts/shared/_base.py

```diff
@@ -51,14 +53,85 @@ def get_diagnostic_filename(basename, cfg, extension='nc'):
     -------
     str:
         A valid path for saving a diagnostic data file.
-
     """
     return os.path.join(
         cfg['work_dir'],
         f"{basename}.{extension}",
     )
 
 
+def save_figure(basename, provenance, cfg, figure=None, close=True, **kwargs):
+    """Save a figure to file.
+
+    Parameters
+    ----------
+    basename: str
+        The basename of the file.
+    provenance: dict
+        The provenance record for the figure.
+    cfg: dict
+        Dictionary with diagnostic configuration.
+    figure: matplotlib.figure.Figure
+        Figure to save.
+    close: bool
+        Close the figure after saving.
+    **kwargs:
+        Keyword arguments to pass to :obj:`matplotlib.figure.Figure.savefig`.
+
+    See Also
+    --------
+    ProvenanceLogger: For an example provenance record that can be used
+        with this function.
+    """
+    if cfg.get('output_file_type') is None:
+        extensions = ('png', 'pdf')
+    elif isinstance(cfg['output_file_type'], str):
+        extensions = (cfg['output_file_type'], )
+    else:
+        extensions = cfg['output_file_type']
+
+    for ext in extensions:
+        filename = Path(cfg['plot_dir']) / ext / f"{basename}.{ext}"
+        filename.parent.mkdir(exist_ok=True)
+        logger.info("Plotting analysis results to %s", filename)
+        fig = plt if figure is None else figure
+        fig.savefig(filename, **kwargs)
```

Comment:
On more thing, but that maybe rather personal preference: Do we want to have a log call here? Something like

```suggestion
        fig.savefig(filename, **kwargs)
        logger.info("Wrote %s", filename)
```

Reply:
Sure, I can add that. It's good practice to always do it in the other order though: first log what you're going to do and then do it, because if it somehow doesn't work out, you can see from the log where things went wrong and if whatever you're going to do takes time the users know what they're waiting for.

Reply:
Oh, I see I misunderstood, there is already a log call for that. Do you prefer to have 2?

Reply:
Ah sorry my bad, didn't see the first log call. Ignore this comment